add cypress-audits to execute lighthouse during end-2-end tests

atomfrede opened the pull request:
Right now the cypress audit uses desktop emulation and no cpu slowdown. With 4x cpu slowdown I get very bad results with angular, so I am not sure what happens in our CI with the current thresholds, especially when the frontend is not packaged in prod mode. Maybe we can add an environment variable to skip audits (e.g. when running in dev mode).

Closes #14615 

---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

<!--
Please also reference the issue number in a commit message to [automatically close the related GitHub issue](https://help.github.com/articles/closing-issues-via-commit-messages/)

Note: It is also possible to add `[skip ci]` or `[ci skip]` to your commit message to skip continuous integration tests
-->

atomfrede commented:
Looks like some configs passes the audit some not (guess somehow depends on the shared runner). I think it would be best to have a dedicated command e.g. npm run audits which just executes the audit spec while the default cypress run does not.

mshima commented on generators/client/templates/angular/package.json.ejs:
```diff
@@ -87,6 +87,8 @@
 <%_ } _%>
 <%_ if (cypressTests) { _%>
     "cypress": "<%= dependabotPackageJson.devDependencies['cypress'] %>",
+    "cypress-audit": "<%= dependabotPackageJson.devDependencies['cypress-audit'] %>",
+    "lighthouse": "<%= dependabotPackageJson.devDependencies['lighthouse'] %>",
```
Inject the dependency like we did at cypress-coverage:
https://github.com/jhipster/generator-jhipster/blob/6b532305ee32497e6a0ed30c3f1140b34d520d99/generators/cypress/index.js#L144-L160

No need to change at every client framework package.json, even blueprints.

mshima commented on generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:
```diff
@@ -38,54 +41,61 @@ module.exports = (on, config) => {
+      fs.writeFileSync('<%= BUILD_DIR %>cypress/lhreport.html', ReportGenerator.generateReport(lighthouseReport.lhr, 'html')); 
+    }),
+    pa11y: pa11y(),
+    <%_ if (authenticationTypeOauth2) { _%>
```
ejs indentation is cleaner if don't follow the generated source indentation.
```suggestion
<%_ if (authenticationTypeOauth2) { _%>
```

mshima commented on generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:
```diff
@@ -38,54 +41,61 @@ module.exports = (on, config) => {
           ignoreHTTPSErrors: true,
           headless: true });
-
+  
```
Some lines with extra blank space.
```suggestion

```

mshima reviewed (commented):
The results seem to change a lot each run.
That may be related to virtual environment shared resources.

So IMO it should be optional and disabled by default.
A test enabling it should be discussed separately, probably an application with no entity with focus on pwa.

mshima commented:
> Looks like some configs passes the audit some not (guess somehow depends on the shared runner). I think it would be best to have a dedicated command e.g. npm run audits which just executes the audit spec while the default cypress run does not.

A custom command would be great too.

atomfrede commented:
> The results seem to change a lot each run.
> That may be related to virtual environment shared resources.
> 
> So IMO it should be optional and disabled by default.
> A test enabling it should be discussed separately, probably an application with no entity with focus on pwa.

I did not expect such results, but yes I will disable it by default and it must be started e.g. with `npm run e2e:audits` or so. Maybe separate the specs also by name would make it easier.

vishal423 commented on generators/cypress/templates/src/test/javascript/cypress/integration/audits.spec.ts.ejs:
```diff
@@ -0,0 +1,46 @@
+-%>
+describe('Audits', () => {
+    beforeEach(() => { 
+      cy.visit('/');
```
It seems that we are trying to calculate the lighthouse score for the application landing page (`/` route). For more comprehensive coverage, we should do that for each route in our application.

atomfrede replied:
Do you expect huge differences between them? But of course that could be extended.

vishal423 replied:
TBH, I don't know. I haven't looked into cypress audits and it's capabilities. Do you have assertions to check on best practices? If yes, then, you may want to run the audit for each page.

In the `svelte` blueprint, I used a different approach and the integration is only in the CI (not in generated code). You can of course do that in the generated code, but that was not the concern that I want to handle at moment.

atomfrede commented:
I have changed it such that 

* audits are in `.audits.ts`
* dedicated cypress config only executing `*.audits.ts` files
* dedicated commands, e.g. `cypress:audits` which opens cypress with the audits configuration
* also updated the readme which did not contain anything about e2e tests when using cypress yet

atomfrede commented on generators/client/templates/angular/package.json.ejs:
```diff
@@ -161,7 +161,7 @@
     "e2e:cypress:headless": "<%= clientPackageManager %> run e2e:cypress -- --headless",
     "e2e:cypress": "cypress run --browser chrome --record ${CYPRESS_ENABLE_RECORD:-false}",
-    "cypress": "cypress open",
+    "cypress": "cypress open"
```
```suggestion
    "cypress": "cypress open",
```

mshima commented on generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:
```diff
@@ -38,54 +41,60 @@ module.exports = (on, config) => {
-
         const page = await browser.newPage();
-
+  
```
```suggestion

```

mshima commented on generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:
```diff
@@ -38,54 +41,60 @@ module.exports = (on, config) => {
           waitUntil: 'networkidle2' // Wait until login-page has been reached
         });
-
+  
```
```suggestion

```

mshima commented on generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:
```diff
@@ -38,54 +41,60 @@ module.exports = (on, config) => {
         await page.type('[name="password"]', password);
         await page.click('input[type="submit"]');
-
+  
```
```suggestion

```

mshima commented on generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:
```diff
@@ -38,54 +41,60 @@ module.exports = (on, config) => {
+  
         await page.waitForNavigation({ waitUntil: 'networkidle2' }); // Wait until redirected back to the app
-
+  
```
```suggestion

```

mshima commented on generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:
```diff
@@ -38,54 +41,60 @@ module.exports = (on, config) => {
+  
         const authCookies = await page.cookies();
-
+  
```
```suggestion

```

mshima commented on generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:
```diff
@@ -38,54 +41,60 @@ module.exports = (on, config) => {
+        return true;
+      })
+    } 
+    <%_ } _%>
```
```suggestion
<%_ } _%>
```

mshima commented on generators/cypress/templates/cypress-audits.json.ejs:
```diff
@@ -0,0 +1,33 @@
+    "viewportHeight": 720,
+    "retries": 2
+  }
+  
```
```suggestion

```

mshima commented on generators/cypress/templates/src/test/javascript/cypress/integration/lighthouse.audits.ts.ejs:
```diff
@@ -0,0 +1,46 @@
+      };
+      cy.lighthouse(customThresholds, desktopConfig);
+    });
+});
```
```suggestion
});

```

mshima reviewed (approved):
I tried to remove whitespaces, but removed lines.

LGTM.

atomfrede commented:
Thanks @mshima for your review. you are awesome!

atomfrede commented:
Bounty claimed https://opencollective.com/generator-jhipster/expenses/40724

pascalgrimaud commented:
@atomfrede : approved
